programs/bpf_loader/src/lib.rs

```diff
@@ -404,94 +459,132 @@ fn process_loader_upgradeable_instruction(
+                ic_logger_msg!(log_collector, "Incorrect buffer authority provided");
+                return Err(InstructionError::IncorrectAuthority);
+            }
+            if !instruction_context.is_signer(
+                instruction_context.get_number_of_program_accounts()
+                    + upgradeable_ins_acc_idx::Write::Authority as usize,
+            )? {
+                ic_logger_msg!(log_collector, "Buffer authority did not sign");
+                return Err(InstructionError::MissingRequiredSignature);
             }
+            let mut program = instruction_context.try_borrow_instruction_account(
+                transaction_context,
+                upgradeable_ins_acc_idx::Write::Buffer as usize,
+            )?;
             write_program_data(
-                first_instruction_account,
+                &mut program,
                 UpgradeableLoaderState::buffer_data_offset()? + offset as usize,
                 &bytes,
-                invoke_context,
+                invoke_context.get_log_collector(),
             )?;
         }
         UpgradeableLoaderInstruction::DeployWithMaxDataLen { max_data_len } => {
-            let payer = keyed_account_at_index(keyed_accounts, first_instruction_account)?;
-            let programdata =
-                keyed_account_at_index(keyed_accounts, first_instruction_account + 1)?;
-            let program = keyed_account_at_index(keyed_accounts, first_instruction_account + 2)?;
-            let buffer = keyed_account_at_index(keyed_accounts, first_instruction_account + 3)?;
-            let rent = get_sysvar_with_account_check::rent(
-                keyed_account_at_index(keyed_accounts, first_instruction_account + 4)?,
+            instruction_context.check_number_of_instruction_accounts(1)?;
+            let programdata_address = *instruction_context.get_instruction_account_key(
+                transaction_context,
+                upgradeable_ins_acc_idx::DeployWithMaxDataLen::ProgramData as usize,
+            )?;
+            let rent = get_sysvar_with_account_check2::rent(
                 invoke_context,
+                instruction_context,
+                upgradeable_ins_acc_idx::DeployWithMaxDataLen::Rent as usize,
             )?;
-            let clock = get_sysvar_with_account_check::clock(
-                keyed_account_at_index(keyed_accounts, first_instruction_account + 5)?,
+            let clock = get_sysvar_with_account_check2::clock(
                 invoke_context,
+                instruction_context,
+                upgradeable_ins_acc_idx::DeployWithMaxDataLen::Clock as usize,
             )?;
-            let authority = keyed_account_at_index(keyed_accounts, first_instruction_account + 7)?;
-            let upgrade_authority_address = Some(*authority.unsigned_key());
-            let upgrade_authority_signer = authority.signer_key().is_none();
+            instruction_context.check_number_of_instruction_accounts(8)?;
```

Comment:
Why call this twice and shouldn't it be 7?

Reply:
It was: `let authority = keyed_account_at_index(keyed_accounts, first_instruction_account + 7)?`

So 8 slots total, no optional ones.
And it is called twice when sysvars are mixed in between as they can throw a different error.